lib/Graph/PlaceholderBindings.cpp

```diff
@@ -150,6 +150,11 @@ Tensor *PlaceholderBindings::allocate(Placeholder *P) {
   DCHECK(!map_.count(P)) << "Placeholder with name \"" << P->getName().str()
                          << "\" already registered";
   Tensor *T = new Tensor(P->getType());
+
+  /// Trainable PH's need to start zeroed.
+  if (P->isTraining()) {
```

Comment:
They shouldn't have to but I suspect you might have been forced to add this to make sure the `gsum` PH of `SGDNode` starts out zeroed.